packages/hash/api/src/model/knowledge/entity.model.ts

```diff
@@ -112,6 +118,145 @@ export default class {
     });
   }
 
+  /**
+   * Create an entity along with any new/existing entities specified through links.
+   *
+   * @param params.createdById the account id that is creating the entity
+   * @param params.entityDefinition the definition of how to get or create the entity optionally with linked entities
+   */
+  static async createEntityWithLinks(
+    graphApi: GraphApi,
+    params: {
+      createdById: string;
+      entityDefinition: KnowledgeEntityDefinition;
+    },
+  ): Promise<EntityModel> {
+    const { createdById, entityDefinition } = params;
+
+    const entitiesInTree = linkedTreeFlatten<
+      KnowledgeEntityDefinition,
+      KnowledgeLinkedEntityDefinition,
+      "linkedEntities",
+      "entity"
+    >(entityDefinition, "linkedEntities", "entity");
+
+    /**
+     * @todo Once the graph API validates the required links of entities on creation, this may have to be reworked in order
+     *   to create valid entities.
+     *   this code currently creates entities first, then links them together.
+     *   See https://app.asana.com/0/1202805690238892/1203046447168478/f
+     */
+    const entities = await Promise.all(
+      entitiesInTree.map(async (definition) => ({
+        link: definition.meta
+          ? {
+              parentIndex: definition.parentIndex,
+              meta: definition.meta,
+            }
+          : undefined,
+        entity: await EntityModel.getOrCreate(graphApi, {
+          createdById,
+          entityDefinition: definition,
+        }),
+      })),
+    );
+
+    await Promise.all(
+      entities.map(async ({ link, entity }) => {
+        if (link) {
+          const parentEntity = entities[link.parentIndex];
+          if (!parentEntity) {
+            throw new ApolloError("Could not find parent entity");
+          }
+          const linkTypeModel = await LinkTypeModel.get(graphApi, {
+            versionedUri: link.meta.linkTypeVersionedUri,
+          });
+
+          // links are created as an outgoing link from the parent entity to the children.
+          await parentEntity.entity.createOutgoingLink(graphApi, {
+            linkTypeModel,
+            targetEntityModel: entity,
+            index: link.meta.index ?? undefined,
+          });
+        }
+      }),
+    );
+
+    if (entities[0]) {
+      // First element will be the root entity.
+      return entities[0].entity;
+    } else {
```

Comment:
I'd move this if check and else above the `Promise.all`.

I know it's functionally the same at the moment because of the `.map`, but if we ever injected some more logic in between it could lead to a subtle bug.

It's also just a lot more readable to throw an error straight away if `entities` is empty. (Not to mention, maybe there is a _slight_ performance overhead to calling `Promise.all`)